Add runtime changes to `2201.5.0` release note (#6589)
ballerina-platform/ballerina-dev-website: downloads/swan-lake-release-notes/swan-lake-2201.5.0 /RELEASE_NOTE.md

- Due to an internal API change, the GraphQL `1.7.0` package is not compatible with older Ballerina versions and older GraphQL versions are not compatible with Ballerina `2201.5.0`. When migrating to Ballerina `2201.5.0` from previous Ballerina distributions, the GraphQL version should be updated to `1.7.0` with this release.

- Fixed a bug that resulted in inconsistent error messages with `cloneWithType` operation.
    
    ```ballerina
    type OpenRecord record {
    };

    public function main() {
        [OpenRecord...] tupleVal = [{"nonJson": xml `<a>abc</a>`}];
        json jsonVal = checkpanic tupleVal.cloneWithType();
    }
    ```

> 

Reviewer: Shall we add the error message as well

Reviewer: @Nadeeshan96

Author: @Nadeeshan96

Reviewer 2: Added the samples.

### New features

#### New Runtime Java APIs

- To clone an `anydata` value with another subtype of `anydata` type, the following API is introduced in the `io.ballerina.runtime.api.utils.ValueUtils` class.
    ```java
    public static Object convert(Object value, Type targetType);
    ```

> - `getFunctionName()` and `getPathParameters()` APIs are introduced in the runtime `io.ballerina.runtime.api.Environment` class. They provide the Ballerina function name and path parameters associated with Java interop methods, respectively.

Reviewer 2: ```suggestion
- `getFunctionName()` and `getPathParameters()` APIs are introduced in the runtime `io.ballerina.runtime.api.Environment` class. They provide the Ballerina function name and path parameters associated with an external Java method, respectively.
```

Reviewer 2: Think it is better this way.

### Improvements

#### Support for command line arguments of built-in subtypes

The command-line arguments of Ballerina built-in subtypes are now supported to be parsed as operands.

For the following `main` function,
```ballerina
public function main(byte byteVal, string:Char charVal, int:Signed8 int8Val) {
}
```
the values can be passed through the command line arguments as follows.
```
bal run -- 1 b 33
```

> #### Support ambiguous union type configurable variables

Reviewer 2: ```suggestion

#### Support ambiguous union type configurable variables
```

When a structural value is provided for a configurable variable of a union type that includes more than one type descriptor, the inherent type used will be the first (leftmost) type descriptor, from which the value can be constructed.

For example, given the following Ballerina code
```ballerina
type Person record {|
    string name;
    string city;
|};

configurable map<string>|Person configVar = ?;
```

> and the `Config.toml` file contains the following,

Reviewer 2: ```suggestion
and the `Config.toml` file contents as follows,
```

```toml
configVar = {name = "Jack", city = "Colombo"}
```
`configVar` is instantiated with the inherent type `map<string>`.

#### Support binding of resource functions to a generic native method

A new way has been introduced to support the binding of any resource function to a generic native method, regardless of the resource path parameters. The generic native method should be defined with a `BArray` parameter, which represents all path parameters. To avoid errors due to overloaded methods, it is recommended to define parameter type constraints as well.

For example, the following Ballerina resource function
```ballerina
isolated resource function get abc/[int p1]/[string p2]/[string p3]/[int ...p4] (string s) = @java:Method {
        'class: "javalibs.app.App",
        name: "getResource",
        paramTypes: ["io.ballerina.runtime.api.values.BObject", "io.ballerina.runtime.api.values.BArray", "io.ballerina.runtime.api.values.BString"]
    } external;
```
can be bound to the following Java method.
```java
public static void getResource(BObject client, BArray path, BString str) {
}
```

#### Improvements in Runtime Java APIs

> - The `getType` runtime API which returns an `ObjectType` in `io.ballerina.runtime.api.values.BObject` class is now deprecated. A new API `getOriginalType` which returns the Type is introduced to return both the `ObjectType` and the type-reference type. We need to modify the previous `getType` API to return the Type after fixing the caching issue [#39850](https://github.com/ballerina-platform/ballerina-lang/issues/39850).

Reviewer 2: ```suggestion
- The `getType` runtime API which returns an `ObjectType` in `io.ballerina.runtime.api.values.BObject` class is now deprecated. A new API `getOriginalType` which returns the `Type` is introduced to return both the `ObjectType` and the type-reference type. We need to modify the previous `getType` API to return the `Type` after fixing the caching issue [#39850](https://github.com/ballerina-platform/ballerina-lang/issues/39850).
```

Reviewer 2: Do we need to include the last sentence? It is like a reminder/to-do for us.

Author: Agreed. 
@warunalakshitha WDYT?

Reviewer: +1

### New features

#### New Runtime Java APIs

> - To clone an `anydata` value with another subtype of `anydata` type, the following API is introduced in the `io.ballerina.runtime.api.utils.ValueUtils` class.

Reviewer 3: ```suggestion
- Introduced the following API in the `io.ballerina.runtime.api.utils.ValueUtils` class to clone an `anydata` value with another subtype of the `anydata` type.
```

    ```java
    public static Object convert(Object value, Type targetType);
    ```

> - `getFunctionName()` and `getPathParameters()` APIs are introduced in the runtime `io.ballerina.runtime.api.Environment` class. They provide the Ballerina function name and path parameters associated with an external Java method, respectively.

Reviewer 3: ```suggestion
- Introduced the `getFunctionName()` and `getPathParameters()` APIs in the runtime `io.ballerina.runtime.api.Environment` class. They provide the Ballerina function name and path parameters associated with an external Java method respectively.
```

### Improvements

> #### Support for command line arguments of built-in subtypes

Reviewer 3: ```suggestion
#### Support for command-line arguments of built-in subtypes
```

> The command-line arguments of Ballerina built-in subtypes are now supported to be parsed as operands.

Reviewer 3: ```suggestion
The command-line arguments of the Ballerina built-in subtypes are now supported to be parsed as operands.
```

For the following `main` function,
```ballerina
public function main(byte byteVal, string:Char charVal, int:Signed8 int8Val) {
}
```

> the values can be passed through the command line arguments as follows.

Reviewer 3: ```suggestion
the values can be passed through command-line arguments as follows.
```

```
bal run -- 1 b 33
```

#### Support ambiguous union type configurable variables

> When a structural value is provided for a configurable variable of a union type that includes more than one type descriptor, the inherent type used will be the first (leftmost) type descriptor, from which the value can be constructed.

Reviewer 3: ```suggestion
When a structural value is provided for a configurable variable of a union type that includes more than one type descriptor, the inherent type used will be the first (leftmost) type descriptor from which the value can be constructed.
```

> For example, given the following Ballerina code

Reviewer 3: ```suggestion
For example, in the following Ballerina code,
```

```ballerina
type Person record {|
    string name;
    string city;
|};

configurable map<string>|Person configVar = ?;
```

> and the `Config.toml` file contents as follows,

Reviewer 3: ```suggestion
and when the `Config.toml` file contents are as follows,
```

```toml
configVar = {name = "Jack", city = "Colombo"}
```

> `configVar` is instantiated with the inherent type `map<string>`.

Reviewer 3: ```suggestion
`configVar` is instantiated with the `map<string>` inherent type.
```

#### Support binding of resource functions to a generic native method

> A new way has been introduced to support the binding of any resource function to a generic native method, regardless of the resource path parameters. The generic native method should be defined with a `BArray` parameter, which represents all path parameters. To avoid errors due to overloaded methods, it is recommended to define parameter type constraints as well.

Reviewer 3: ```suggestion
A new way has been introduced to support the binding of any resource function to a generic native method regardless of the resource path parameters. The generic native method should be defined with a `BArray` parameter, which represents all the path parameters. To avoid errors due to overloaded methods, it is recommended to define the parameter type constraints as well.
```

For example, the following Ballerina resource function
```ballerina
isolated resource function get abc/[int p1]/[string p2]/[string p3]/[int ...p4] (string s) = @java:Method {
        'class: "javalibs.app.App",
        name: "getResource",
        paramTypes: ["io.ballerina.runtime.api.values.BObject", "io.ballerina.runtime.api.values.BArray", "io.ballerina.runtime.api.values.BString"]
    } external;
```
can be bound to the following Java method.
```java
public static void getResource(BObject client, BArray path, BString str) {
}
```

#### Improvements in Runtime Java APIs

> - The `getType` runtime API which returns an `ObjectType` in `io.ballerina.runtime.api.values.BObject` class is now deprecated. A new API `getOriginalType` which returns the `Type` is introduced to return both the `ObjectType` and the type-reference type. We need to modify the previous `getType` API to return the `Type` after fixing the caching issue [#39850](https://github.com/ballerina-platform/ballerina-lang/issues/39850).

Reviewer 3: ```suggestion
- The `getType` runtime API, which returns an `ObjectType` in the `io.ballerina.runtime.api.values.BObject` class is now deprecated. A new `getOriginalType` API, which returns the `Type` is introduced to return both the `ObjectType` and the type-reference type. You need to modify the previous `getType` API to return the `Type` after fixing the [caching issue](https://github.com/ballerina-platform/ballerina-lang/issues/39850).
```

> - The java constant `XMLNS` in `io.ballerina.runtime.api.values.BXmlItem` runtime class is now deprecated. The constant `javax.xml.XMLConstants.XMLNS_ATTRIBUTE` needs to be used instead.

Reviewer 3: ```suggestion
- The `XMLNS` Java constant in the `io.ballerina.runtime.api.values.BXmlItem` runtime class is now deprecated. Instead, the `javax.xml.XMLConstants.XMLNS_ATTRIBUTE` constant needs to be used.
```
